torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py

```diff
@@ -0,0 +1,195 @@
+        # as a regular key-value dictionary.
+        #
+        # As a workaround we initially set a sentinel value as the rendezvous
+        # state. Whenever this value gets returned we treat it as a None.
+        self._call_store("compare_set", self._key, "", self._NULL_SENTINEL)
+
+    @property
+    def name(self) -> str:
+        """See base class."""
+        return "c10d-experimental"
+
+    @property
+    def store(self) -> Store:
+        """Gets the `Store` instance used to communicate with the C10d store."""
+        return self._store
+
+    @property
+    def key(self) -> str:
+        """Gets the key under which the rendezvous state is stored."""
+        return self._key
+
+    def get_state(self) -> Optional[Tuple[bytes, Token]]:
+        """See base class."""
+        base64_state: bytes = self._call_store("get", self._key)
+
+        return self._decode_state(base64_state)
+
+    def set_state(
+        self, state: bytes, token: Optional[Token] = None
+    ) -> Optional[Tuple[bytes, Token]]:
+        """See base class."""
+        base64_state_str: str = codecs.encode(state, "base64").decode()
+
+        if token:
+            # Shortcut if we know for sure that the token is not valid.
+            if not isinstance(token, bytes):
+                return self.get_state()
+
+            token = token.decode()
+        else:
+            token = self._NULL_SENTINEL
+
+        base64_state: bytes = self._call_store("compare_set", self._key, token, base64_state_str)
+
+        return self._decode_state(base64_state)
+
+    def _call_store(self, store_op: str, *args, **kwargs) -> Any:
+        try:
+            return getattr(self._store, store_op)(*args, **kwargs)
+        except (ValueError, RuntimeError) as exc:
+            raise RendezvousConnectionError(
+                "The connection to the C10d store has failed. See inner exception for details."
+            ) from exc
+
+    def _decode_state(self, base64_state: bytes) -> Optional[Tuple[bytes, Token]]:
+        if base64_state == self._NULL_SENTINEL.encode():
+            return None
+        try:
+            state = codecs.decode(base64_state, "base64")
+        except binascii.Error as exc:
+            raise RendezvousStateError(
+                "The state object is corrupt. See inner exception for details."
+            ) from exc
+
+        return state, base64_state
+
+
+def _create_tcp_store(params: RendezvousParameters) -> TCPStore:
+    host, port = _parse_rendezvous_endpoint(params.endpoint, default_port=29500)
+
+    cfg_is_host = params.get_as_bool("is_host")
+    # If the user has explicitly specified whether our process should host the
+    # the store, respect it.
+    if cfg_is_host is not None:
+        is_host = cfg_is_host
+    # Otherwise try to determine whether we are the host based on our hostname
+    # and IP address.
+    else:
+        is_host = _matches_machine_hostname(host)
```

Comment:
Would it be simpler if just required `is_host` to be always specified and error out if cfg_is_host is None?

Reply:
It would be definitely simpler on our end, but it would also be slightly inconvenient and redundant for the user. In normal circumstances (except for the edge case I mentioned above where you might have more than one process on the same machine) the endpoint already conveys all the information you need to determine whether you are the host or not. If the specified endpoint's hostname equals your machine's hostname, you know that you will be the host. Also this way you can have a single launcher script that can be used by all machines identically. If `is_host` would be mandatory, you would have to make sure that there are two versions of the very same launcher script; one for the machine that corresponds to the endpoint with `is_host=True`, and one for other machines with `is_host=False`.